### ISSUE TITLE
ServerFarmId needs to be fully qualified to show as unchanged

### ISSUE THREAD

**Issue by isaacabraham** [Azure/arm-template-whatif#123]
## Describe the noise

**Microsoft.Web/sites**

**2016-08-01**

**Azure CLI**

```bash
~ properties.serverFarmId: "subscriptions/<subId>/resourceGroups/<resource-group-id>/providers/Microsoft.Web/serverfarms/<server-farm-id>" => "<server-farm-id>"
```

Deploying a web site, unless you supply the fully qualified path of the `serverFarmId` in the ARM template, what-if shows this as a change per the example above. I would expect nothing to do changed since the template has not changed.

**Comment by alex-frankel** [comment]
Isaac, does this only work if the serverFarm is in the same sub & rg as the website? TBH, I'm surprised they accept the "shorthand" version of the serverFarmId. Most people use the `resourceId()` function when wiring things up like this FWIW.

**Comment by isaacabraham** [comment]
@alex-frankel yep. From a Farmer perspective 99% of the time that's exactly what happens, because the app service and server farm are logically treated as one. The nice thing about just using the resource name is that the ARM template itself is somewhat more readable (and if you're manually authoring / modifying, it's quicker to do).

However, I'm keen to adhere to best practices - when using `resourceId()` should we be fully qualifying up to the resource group level, or even further e.g. subscription id (this seems overkill to me when I think about it)?

**Comment by alex-frankel** [comment]
I think it's fine to just do:

```json
"serverFarmId": "[resourceId('Microsoft.Web/serverFarms', '<server-farm-name>')] 
```

If you know it's in the same sub and rg, you can leave it as the above. If you are not sure, then you can qualify with rg and sub as needed. If the user ever needs to edit the ARM JSON, resourceId is a pretty well known pattern so user's should be able to make the adjustment.

Thoughts?

**Comment by isaacabraham** [comment]
Just to clarify though - in Farmer you *can* link a Farmer-created webapp to what we call an "unmanaged" farm i.e. a server farm that's in another resource group, created outside of Farmer. In such a case, you can manually pass a fully-qualified resourceId and we use that instead.

**Comment by isaacabraham** [comment]
If we include only the type, what's the real benefit of that - that it prevents name clashes of two resources of different types but the same name (does that ever happen? no-one has ever complained in Farmer about this limitation...)?

**Comment by alex-frankel** [comment]
My feeling is that encouraging people to use the resourceId() pattern is a good thing, whereas just using the short name may work fine now, but leaves more for you to learn later. That's not a strong opinion btw.

Otherwise, you are right that the only benefit is cases where you have the same name and a different resource type. As you note, we don't see often.

**Comment by torreymicrosoft** [comment]
Thanks for your patience, Isaac, this was recently picked up as one of the older open issues for the ARM Deployments team to revisit. Sorry for the long wait.

What is happening is a plain string comparison. Your template supplies `properties.serverFarmId` as the plan short name, the Microsoft.Web resource provider stores it expanded to the full `/subscriptions/.../providers/Microsoft.Web/serverfarms/<name>` form, and what-if compares the two strings and calls it a change. I checked the current `Microsoft.Web/sites` type (2024-11-01) and a bare short name is still perfectly legal there, so your template was never the problem. This repo labelled it `normalization`, which is exactly right.

Two things now.

Alex's suggestion on the thread still stands as the immediate fix: use `[resourceId('Microsoft.Web/serverFarms', '<server-farm-name>')]`, and only qualify with resource group or subscription when the plan lives elsewhere. Your point about Farmer treating the site and the plan as one unit is fair, and it is why the short name felt natural.

Beyond that, what-if for Azure Deployment Stacks went generally available today, with noise reduction enabled in all regions. A stack records a what-if baseline when it is deployed or updated, and on later runs any property that is unchanged between the baseline and the current evaluation is removed from the result. A short name versus its expanded ID is the same difference on every run, so `serverFarmId` stops being reported.

For a resource group scope stack on Azure CLI:

```
az stack-whatif group create \
  --name site-preview \
  --resource-group <your-rg> \
  --stack-id <deployment-stack-resource-id> \
  --template-file main.bicep \
  --action-on-unmanage detachAll \
  --deny-settings-mode none \
  --retention-interval PT3H
```

Announcement: https://techcommunity.microsoft.com/blog/azuregovernanceandmanagementblog/now-generally-available-what-if-for-azure-deployment-stacks/4547614
Docs: https://learn.microsoft.com/azure/azure-resource-manager/bicep/deployment-stacks-what-if

The baseline is recorded at stack deploy or update time, so deploy or update the stack once to establish it, and noise reduction does not remove every difference.

I am closing this out. If `serverFarmId` still shows as a change through stacks what-if, reopen here or file it in `Azure/deployment-stacks` with the result resource ID.
